channelz: stage 4 - add security and socket option info

lyuxuan opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/grpc/grpc-go/1965)
<!-- Reviewable:end -->

menghanl commented:
Reviewed 10 of 20 files at r1, 4 of 7 files at r2.
Review status: 14 of 18 files reviewed at latest revision, all discussions resolved.

---

*[channelz/types.go, line 423 at r2](https://reviewable.io/reviews/grpc/grpc-go/1965#-LBY3drnAHGpCeUTIp4b:-LBY3drnAHGpCeUTIp4c:b-jezj1i) ([raw file](https://github.com/grpc/grpc-go/blob/686bbac22514f2fb21aacbd96abfa3ef25e6faac/channelz/types.go#L423)):*
> ```Go
> // Security defines the interface that security protocols should implement in order
> // to provide security info to channelz.
> type Security interface {
> ```

Move this to a sub package?

`channelz` package defines the API between the clientconn and the channelz store.
This is more like a helper function for the clientconn to get info from creds. It's not part of the API.

---

*[channelz/types_linux.go, line 24 at r2](https://reviewable.io/reviews/grpc/grpc-go/1965#-LBYMfpl5CfFdEADluNq:-LBYMfpl5CfFdEADluNr:b-8tro9i) ([raw file](https://github.com/grpc/grpc-go/blob/686bbac22514f2fb21aacbd96abfa3ef25e6faac/channelz/types_linux.go#L24)):*
> ```Go
> 	"syscall"
> 
> 	"golang.org/x/sys/unix"
> ```

Will this import cause the build to fail on non-linux? Or just function calls will return error?

---

*[channelz/types_linux.go, line 30 at r2](https://reviewable.io/reviews/grpc/grpc-go/1965#-LBYMsfx2dFHkjF1HVIt:-LBYMsfx2dFHkjF1HVIu:b-6jp4sq) ([raw file](https://github.com/grpc/grpc-go/blob/686bbac22514f2fb21aacbd96abfa3ef25e6faac/channelz/types_linux.go#L30)):*
> ```Go
> // getter function to obtain info from fd.
> type SocketOptionData struct {
> 	Linger      *unix.Linger
> ```

This field is not defined in `_windows.go`, but is referenced in other files (e.g. `service.go`), even on windows.

---

*[channelz/util_default.go, line 1 at r2](https://reviewable.io/reviews/grpc/grpc-go/1965#-LBYOE755qHnakwmrXFt:-LBYOE755qHnakwmrXFu:b-u61l2l) ([raw file](https://github.com/grpc/grpc-go/blob/686bbac22514f2fb21aacbd96abfa3ef25e6faac/channelz/util_default.go#L1)):*
> ```Go
> // +build !darwin,!dragonfly,!freebsd,!linux,!netbsd,!openbsd,!solaris !go1.9
> ```

Rename this file. `default` sounds too normal.

---

*[channelz/util_unix_go19.go, line 1 at r2](https://reviewable.io/reviews/grpc/grpc-go/1965#-LBYO29Z4Rtm0ZtMSEXg:-LBYO29Z4Rtm0ZtMSEXh:b5hb6n2) ([raw file](https://github.com/grpc/grpc-go/blob/686bbac22514f2fb21aacbd96abfa3ef25e6faac/channelz/util_unix_go19.go#L1)):*
> ```Go
> // +build darwin dragonfly freebsd linux netbsd openbsd solaris
> ```

Why do we need this build constraint?
And why is this not just windows or not?

---

*[channelz/util_unix_go19.go, line 29 at r2](https://reviewable.io/reviews/grpc/grpc-go/1965#-LBYBFzvEqLYL04sOAMi:-LBYBFzwCvF3OXJwAvMU:b-jqcfzd) ([raw file](https://github.com/grpc/grpc-go/blob/686bbac22514f2fb21aacbd96abfa3ef25e6faac/channelz/util_unix_go19.go#L29)):*
> ```Go
> 
> // GetSocketOption gets the socket option info of the conn.
> func GetSocketOption(socket interface{}) *SocketOptionData {
> ```

This is also a helper function, not part of the API.
In other words, this is not necessary for channelz, the clientconn could generate the same result without calling this function.

---

*[channelz/service/service.go, line 38 at r2](https://reviewable.io/reviews/grpc/grpc-go/1965#-LBYOkKhCbOt1CcuAJQ5:-LBYOkKhCbOt1CcuAJQ6:b-7235sw) ([raw file](https://github.com/grpc/grpc-go/blob/686bbac22514f2fb21aacbd96abfa3ef25e6faac/channelz/service/service.go#L38)):*
> ```Go
> 
> const (
> 	secToNano  = 1e9
> ```

Make a function to do the time conversion.

---

*[channelz/service/service.go, line 145 at r2](https://reviewable.io/reviews/grpc/grpc-go/1965#-LBY2esU0nSW3ah5fesN:-LBY2esU0nSW3ah5fesO:b-76rf9k) ([raw file](https://github.com/grpc/grpc-go/blob/686bbac22514f2fb21aacbd96abfa3ef25e6faac/channelz/service/service.go#L145)):*
> ```Go
> 			LocalCertificate:  v.LocalCertificate,
> 			RemoteCertificate: v.RemoteCertificate,
> 		},
> ```

Move `}`'s to one line

---

*[credentials/credentials.go, line 35 at r2](https://reviewable.io/reviews/grpc/grpc-go/1965#-LBYQrHc8e_HhwZkO8rF:-LBYQrHc8e_HhwZkO8rG:bjd9k9q) ([raw file](https://github.com/grpc/grpc-go/blob/686bbac22514f2fb21aacbd96abfa3ef25e6faac/credentials/credentials.go#L35)):*
> ```Go
> 
> 	"google.golang.org/grpc/channelz"
> ```

Delete blank line.

---

*[credentials/credentials.go, line 172 at r2](https://reviewable.io/reviews/grpc/grpc-go/1965#-LBYM7hi8p-CrFWLeTrL:-LBYM7hi8p-CrFWLeTrM:bl2h66y) ([raw file](https://github.com/grpc/grpc-go/blob/686bbac22514f2fb21aacbd96abfa3ef25e6faac/credentials/credentials.go#L172)):*
> ```Go
> 		return nil, nil, ctx.Err()
> 	}
> 	return tlsConn{conn, rawConn}, TLSInfo{conn.ConnectionState()}, nil
> ```

Names of the fields.

---

*[credentials/credentials_util_go19.go, line 5 at r2](https://reviewable.io/reviews/grpc/grpc-go/1965#-LBYRJP81Sj8LkT3a2IK:-LBYRJP81Sj8LkT3a2IL:bb6ht2i) ([raw file](https://github.com/grpc/grpc-go/blob/686bbac22514f2fb21aacbd96abfa3ef25e6faac/credentials/credentials_util_go19.go#L5)):*
> ```Go
> /*
>  *
>  * Copyright 2014 gRPC authors.
> ```

2018

---

*[credentials/credentials_util_go19.go, line 36 at r2](https://reviewable.io/reviews/grpc/grpc-go/1965#-LBYSVLc70yzBXcwdpB5:-LBYSVLc70yzBXcwdpB6:bq1esnr) ([raw file](https://github.com/grpc/grpc-go/blob/686bbac22514f2fb21aacbd96abfa3ef25e6faac/credentials/credentials_util_go19.go#L36)):*
> ```Go
> 
> func (c tlsConn) SyscallConn() (syscall.RawConn, error) {
> 	conn, ok := c.rawConn.(syscall.Conn)
> ```

Since we only care about method `SyscallConn`, do
```go
rawConn.(interface{
  SyscallConn() (syscall.RawConn, error)
})
```

So we don't need the `_pre19.go` file for this.

But this works only if `syscall.RawConn` is defined pre 1.9.

---

*[credentials/credentials_util_pre_go19.go, line 5 at r2](https://reviewable.io/reviews/grpc/grpc-go/1965#-LBYTK5xAmVDVhpeWZDL:-LBYTK5xAmVDVhpeWZDM:bb6ht2i) ([raw file](https://github.com/grpc/grpc-go/blob/686bbac22514f2fb21aacbd96abfa3ef25e6faac/credentials/credentials_util_pre_go19.go#L5)):*
> ```Go
> /*
>  *
>  * Copyright 2014 gRPC authors.
> ```

2018

---

*[credentials/credentials_util_pre_go19.go, line 30 at r2](https://reviewable.io/reviews/grpc/grpc-go/1965#-LBYT7r60saglQCZJW9H:-LBYT7r60saglQCZJW9I:b78y6qi) ([raw file](https://github.com/grpc/grpc-go/blob/686bbac22514f2fb21aacbd96abfa3ef25e6faac/credentials/credentials_util_pre_go19.go#L30)):*
> ```Go
> type tlsConn struct {
> 	*tls.Conn
> 	rawConn net.Conn
> ```

Why keeping the rawConn?

---

*[test/channelz_test.go, line 125 at r2](https://reviewable.io/reviews/grpc/grpc-go/1965#-LBYTO_P1bVLOI38ELVT:-LBYTO_P1bVLOI38ELVU:bcq0qt5) ([raw file](https://github.com/grpc/grpc-go/blob/686bbac22514f2fb21aacbd96abfa3ef25e6faac/test/channelz_test.go#L125)):*
> ```Go
> 			ccs = append(ccs, cc)
> 		}
> ```

Revert this.

---


*Comments from [Reviewable](https://reviewable.io/reviews/grpc/grpc-go/1965)*
<!-- Sent from Reviewable.io -->

lyuxuan commented:
Review status: 3 of 17 files reviewed at latest revision, 15 unresolved discussions.

---

*[channelz/types.go, line 423 at r2](https://reviewable.io/reviews/grpc/grpc-go/1965#-LBY3drnAHGpCeUTIp4b:-LCG9Ma7CdzruWiLPWYO:bc1avsv) ([raw file](https://github.com/grpc/grpc-go/blob/686bbac22514f2fb21aacbd96abfa3ef25e6faac/channelz/types.go#L423)):*
<details><summary><i>Previously, menghanl (Menghan Li) wrote…</i></summary><blockquote>

Move this to a sub package?

`channelz` package defines the API between the clientconn and the channelz store.
This is more like a helper function for the clientconn to get info from creds. It's not part of the API.
</blockquote></details>

moved to credentials package.

---

*[channelz/util_unix_go19.go, line 1 at r2](https://reviewable.io/reviews/grpc/grpc-go/1965#-LBYO29Z4Rtm0ZtMSEXg:-LCGc9aj4Zjdvmka_Xgf:b-d6d9fr) ([raw file](https://github.com/grpc/grpc-go/blob/686bbac22514f2fb21aacbd96abfa3ef25e6faac/channelz/util_unix_go19.go#L1)):*
<details><summary><i>Previously, menghanl (Menghan Li) wrote…</i></summary><blockquote>

Why do we need this build constraint?
And why is this not just windows or not?
</blockquote></details>

https://github.com/golang/sys/blob/master/unix/syscall_unix.go contains definition for function like GetsockoptLinger, GetsockoptTimeval, and it has the build tag listed here. Therefore, I have to have the same build tag here. FYI, I've made some changes to the build tag.

---

*[channelz/util_unix_go19.go, line 29 at r2](https://reviewable.io/reviews/grpc/grpc-go/1965#-LBYBFzvEqLYL04sOAMi:-LCGuVZ64uFDn9y7LEwx:bf0iy99) ([raw file](https://github.com/grpc/grpc-go/blob/686bbac22514f2fb21aacbd96abfa3ef25e6faac/channelz/util_unix_go19.go#L29)):*
<details><summary><i>Previously, menghanl (Menghan Li) wrote…</i></summary><blockquote>

This is also a helper function, not part of the API.
In other words, this is not necessary for channelz, the clientconn could generate the same result without calling this function.
</blockquote></details>

If we move this function to transport package, the it will complicate the package by adding two more files for different go version.

---

*[channelz/service/service.go, line 38 at r2](https://reviewable.io/reviews/grpc/grpc-go/1965#-LBYOkKhCbOt1CcuAJQ5:-LCGsl-i5X7N3h2wm0l7:bb7f2mf) ([raw file](https://github.com/grpc/grpc-go/blob/686bbac22514f2fb21aacbd96abfa3ef25e6faac/channelz/service/service.go#L38)):*
<details><summary><i>Previously, menghanl (Menghan Li) wrote…</i></summary><blockquote>

Make a function to do the time conversion.
</blockquote></details>

done

---

*[channelz/service/service.go, line 145 at r2](https://reviewable.io/reviews/grpc/grpc-go/1965#-LBY2esU0nSW3ah5fesN:-LCGslj5Edch_YG5t7om:bb7f2mf) ([raw file](https://github.com/grpc/grpc-go/blob/686bbac22514f2fb21aacbd96abfa3ef25e6faac/channelz/service/service.go#L145)):*
<details><summary><i>Previously, menghanl (Menghan Li) wrote…</i></summary><blockquote>

Move `}`'s to one line
</blockquote></details>

done

---

*[credentials/credentials.go, line 35 at r2](https://reviewable.io/reviews/grpc/grpc-go/1965#-LBYQrHc8e_HhwZkO8rF:-LCGnaN-CTbzK5PipZGb:bb7f2mf) ([raw file](https://github.com/grpc/grpc-go/blob/686bbac22514f2fb21aacbd96abfa3ef25e6faac/credentials/credentials.go#L35)):*
<details><summary><i>Previously, menghanl (Menghan Li) wrote…</i></summary><blockquote>

Delete blank line.
</blockquote></details>

done

---

*[credentials/credentials.go, line 172 at r2](https://reviewable.io/reviews/grpc/grpc-go/1965#-LBYM7hi8p-CrFWLeTrL:-LCGobiyENLdKDdDekYy:bb7f2mf) ([raw file](https://github.com/grpc/grpc-go/blob/686bbac22514f2fb21aacbd96abfa3ef25e6faac/credentials/credentials.go#L172)):*
<details><summary><i>Previously, menghanl (Menghan Li) wrote…</i></summary><blockquote>

Names of the fields.
</blockquote></details>

done

---

*[credentials/credentials_util_go19.go, line 5 at r2](https://reviewable.io/reviews/grpc/grpc-go/1965#-LBYRJP81Sj8LkT3a2IK:-LCGmoypB-MiWuzVakkv:bb7f2mf) ([raw file](https://github.com/grpc/grpc-go/blob/686bbac22514f2fb21aacbd96abfa3ef25e6faac/credentials/credentials_util_go19.go#L5)):*
<details><summary><i>Previously, menghanl (Menghan Li) wrote…</i></summary><blockquote>

2018
</blockquote></details>

done

---

*[credentials/credentials_util_go19.go, line 36 at r2](https://reviewable.io/reviews/grpc/grpc-go/1965#-LBYSVLc70yzBXcwdpB5:-LCGmqj0EQ1Z8kUUpy9m:b-qpmv3v) ([raw file](https://github.com/grpc/grpc-go/blob/686bbac22514f2fb21aacbd96abfa3ef25e6faac/credentials/credentials_util_go19.go#L36)):*
<details><summary><i>Previously, menghanl (Menghan Li) wrote…</i></summary><blockquote>

Since we only care about method `SyscallConn`, do
```go
rawConn.(interface{
  SyscallConn() (syscall.RawConn, error)
})
```

So we don't need the `_pre19.go` file for this.

But this works only if `syscall.RawConn` is defined pre 1.9.
</blockquote></details>

Unfortunately, syscall.RawConn is not defined pre 1.9

---

*[credentials/credentials_util_pre_go19.go, line 5 at r2](https://reviewable.io/reviews/grpc/grpc-go/1965#-LBYTK5xAmVDVhpeWZDL:-LCGn0S6EWXXzg-Nl3og:bb7f2mf) ([raw file](https://github.com/grpc/grpc-go/blob/686bbac22514f2fb21aacbd96abfa3ef25e6faac/credentials/credentials_util_pre_go19.go#L5)):*
<details><summary><i>Previously, menghanl (Menghan Li) wrote…</i></summary><blockquote>

2018
</blockquote></details>

done

---

*[credentials/credentials_util_pre_go19.go, line 30 at r2](https://reviewable.io/reviews/grpc/grpc-go/1965#-LBYT7r60saglQCZJW9H:-LCGn1vE-4XZ_Z3XQPtA:b-rj42rt) ([raw file](https://github.com/grpc/grpc-go/blob/686bbac22514f2fb21aacbd96abfa3ef25e6faac/credentials/credentials_util_pre_go19.go#L30)):*
<details><summary><i>Previously, menghanl (Menghan Li) wrote…</i></summary><blockquote>

Why keeping the rawConn?
</blockquote></details>

That's how we get the socket options, i.e. by doing the SyscallConn(rawConn) and then Control()

---

*[channelz/types_linux.go, line 24 at r2](https://reviewable.io/reviews/grpc/grpc-go/1965#-LBYMfpl5CfFdEADluNq:-LCGYLUn-CjXJiA2Gmgb:b-yomekc) ([raw file](https://github.com/grpc/grpc-go/blob/686bbac22514f2fb21aacbd96abfa3ef25e6faac/channelz/types_linux.go#L24)):*
<details><summary><i>Previously, menghanl (Menghan Li) wrote…</i></summary><blockquote>

Will this import cause the build to fail on non-linux? Or just function calls will return error?
</blockquote></details>

nice catch!

---

*[channelz/types_linux.go, line 30 at r2](https://reviewable.io/reviews/grpc/grpc-go/1965#-LBYMsfx2dFHkjF1HVIt:-LCGYCqGEQnM6ldxhp0X:b-en2s5o) ([raw file](https://github.com/grpc/grpc-go/blob/686bbac22514f2fb21aacbd96abfa3ef25e6faac/channelz/types_linux.go#L30)):*
<details><summary><i>Previously, menghanl (Menghan Li) wrote…</i></summary><blockquote>

This field is not defined in `_windows.go`, but is referenced in other files (e.g. `service.go`), even on windows.
</blockquote></details>

I added some dummy struct definitions for SocketOptionData for non unix build

---

*[channelz/util_default.go, line 1 at r2](https://reviewable.io/reviews/grpc/grpc-go/1965#-LBYOE755qHnakwmrXFt:-LCG_XRB-u6lRIx5x_2_:b-y3v87) ([raw file](https://github.com/grpc/grpc-go/blob/686bbac22514f2fb21aacbd96abfa3ef25e6faac/channelz/util_default.go#L1)):*
<details><summary><i>Previously, menghanl (Menghan Li) wrote…</i></summary><blockquote>

Rename this file. `default` sounds too normal.
</blockquote></details>

renamed to util_nonlinux_pre_go19.go

---

*[test/channelz_test.go, line 125 at r2](https://reviewable.io/reviews/grpc/grpc-go/1965#-LBYTO_P1bVLOI38ELVT:-LCGnNDg-xF-vH7IG8sE:b32wtd) ([raw file](https://github.com/grpc/grpc-go/blob/686bbac22514f2fb21aacbd96abfa3ef25e6faac/test/channelz_test.go#L125)):*
<details><summary><i>Previously, menghanl (Menghan Li) wrote…</i></summary><blockquote>

Revert this.
</blockquote></details>

ok

---


*Comments from [Reviewable](https://reviewable.io/reviews/grpc/grpc-go/1965)*
<!-- Sent from Reviewable.io -->

menghanl commented:
Reviewed 1 of 23 files at r1, 3 of 19 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[channelz/util_test.go, line 31 at r3](https://reviewable.io/reviews/grpc/grpc-go/1965#-LCoyDlF-bPD4sXmu54f:-LCoyDlF-bPD4sXmu54g:b-e0o0fx) ([raw file](https://github.com/grpc/grpc-go/blob/a628f14e54172a5bdc9dda43db4237452f0a3685/channelz/util_test.go#L31)):*
> ```Go
> 
> 	"google.golang.org/grpc/channelz"
> ```

Remove this blank line.

---

*[channelz/util_test.go, line 65 at r3](https://reviewable.io/reviews/grpc/grpc-go/1965#-LCp-Nff7CpsCG2DucD4:-LCp-Nff7CpsCG2DucD5:b-pi1vbx) ([raw file](https://github.com/grpc/grpc-go/blob/a628f14e54172a5bdc9dda43db4237452f0a3685/channelz/util_test.go#L65)):*
> ```Go
> 		t.Fatalf("net.Listen(%s,%s) failed with err: %v", network, addr, err)
> 	}
> 	done := startServer(ln)
> ```

something is wrong with this `startServer` function

---

*[channelz/util_test.go, line 73 at r3](https://reviewable.io/reviews/grpc/grpc-go/1965#-LCoyef06tjGBxUMYg1U:-LCoyef06tjGBxUMYg1V:btqtve9) ([raw file](https://github.com/grpc/grpc-go/blob/a628f14e54172a5bdc9dda43db4237452f0a3685/channelz/util_test.go#L73)):*
> ```Go
> 	}
> 
> 	l := &unix.Linger{Onoff: 1, Linger: 5}
> ```

The content of this struct will be modified by the syscall later, right?
Just leave it empty? Or use `new`.

---

*[channelz/util_test.go, line 110 at r3](https://reviewable.io/reviews/grpc/grpc-go/1965#-LCozvUMCxPfZiSYKw_h:-LCozvUMCxPfZiSYKw_i:b-7rrkab) ([raw file](https://github.com/grpc/grpc-go/blob/a628f14e54172a5bdc9dda43db4237452f0a3685/channelz/util_test.go#L110)):*
> ```Go
> 	}
> 
> 	ln.Close()
> ```

defer

---

*[channelz/service/service.go, line 153 at r3](https://reviewable.io/reviews/grpc/grpc-go/1965#-LCp-cso9bN7DD_9izAJ:-LCp-cso9bN7DD_9izAK:b7atwln) ([raw file](https://github.com/grpc/grpc-go/blob/a628f14e54172a5bdc9dda43db4237452f0a3685/channelz/service/service.go#L153)):*
> ```Go
> 		anyval, err := ptypes.MarshalAny(v.Value)
> 		if err != nil {
> 			return &pb.Security{Model: &pb.Security_Other{Other: &pb.Security_OtherSecurity{
> ```

Move this code around so you create one local variable and only set its field when error is not nil.

---

*[channelz/service/service.go, line 168 at r3](https://reviewable.io/reviews/grpc/grpc-go/1965#-LCow4sY4raX9G6AX-mM:-LCow4sY4raX9G6AX-mN:b-p90iq7) ([raw file](https://github.com/grpc/grpc-go/blob/a628f14e54172a5bdc9dda43db4237452f0a3685/channelz/service/service.go#L168)):*
> ```Go
> 	var opts []*pb.SocketOption
> 	if skopts.Linger != nil {
> 		additional, err := ptypes.MarshalAny(&pb.SocketOptionLinger{Active: skopts.Linger.Onoff != 0, Duration: ptypes.DurationProto(time.Duration(secToNano(int64(skopts.Linger.Linger))))})
> ```

Move this and other code that accesses internal fields of SocketOptionData to _unix.go and _nonunix.go files. So you won't need the dummy non-unix types, either.

---

*[channelz/service/service.go, line 168 at r3](https://reviewable.io/reviews/grpc/grpc-go/1965#-LCp1WHL5zNTsJSnhDBM:-LCp1WHL5zNTsJSnhDBN:b-8ztq8q) ([raw file](https://github.com/grpc/grpc-go/blob/a628f14e54172a5bdc9dda43db4237452f0a3685/channelz/service/service.go#L168)):*
> ```Go
> 	var opts []*pb.SocketOption
> 	if skopts.Linger != nil {
> 		additional, err := ptypes.MarshalAny(&pb.SocketOptionLinger{Active: skopts.Linger.Onoff != 0, Duration: ptypes.DurationProto(time.Duration(secToNano(int64(skopts.Linger.Linger))))})
> ```

And put the fields of the struct to multiple lines.

---

*[channelz/service/service_test.go, line 35 at r3](https://reviewable.io/reviews/grpc/grpc-go/1965#-LCp2V5R0AiCilGwflS1:-LCp2V5R0AiCilGwflS2:b-gi79kz) ([raw file](https://github.com/grpc/grpc-go/blob/a628f14e54172a5bdc9dda43db4237452f0a3685/channelz/service/service_test.go#L35)):*
> ```Go
> 	"golang.org/x/sys/unix"
> 	"google.golang.org/grpc/channelz"
> 	pb "google.golang.org/grpc/channelz/service_proto"
> ```

channelzpb "google.golang.org/grpc/channelz/"

---

*[channelz/service/service_test.go, line 41 at r3](https://reviewable.io/reviews/grpc/grpc-go/1965#-LCp4Sx79HuTsqYYmZsp:-LCp4Sx79HuTsqYYmZsq:bb6tdel) ([raw file](https://github.com/grpc/grpc-go/blob/a628f14e54172a5bdc9dda43db4237452f0a3685/channelz/service/service_test.go#L41)):*
> ```Go
> 
> func init() {
> 	proto.RegisterType((*OtherSecurityValue)(nil), "grpc.credentials.OtherSecurityValue")
> ```

Move this next to the definition of `OtherSecurityValue`, and add a comment that it's needed by `UnmarshalAny`

---

*[channelz/service/service_test.go, line 180 at r3](https://reviewable.io/reviews/grpc/grpc-go/1965#-LCp10KD4-0G6fx0UjFu:-LCp10KD4-0G6fx0UjFv:bri3iog) ([raw file](https://github.com/grpc/grpc-go/blob/a628f14e54172a5bdc9dda43db4237452f0a3685/channelz/service/service_test.go#L180)):*
> ```Go
> 	if protoDuration := protoLinger.GetDuration(); protoDuration != nil {
> 		if dur, err := ptypes.Duration(protoDuration); err == nil {
> 			linger.Linger = int32(int64(dur) / 1e9)
> ```

Similar to previous comments on making a function to convert duration, this and the following `1e3`, `1e9` could also be a function.

---

*[credentials/credentials_util_go19.go, line 35 at r3](https://reviewable.io/reviews/grpc/grpc-go/1965#-LCp0LZk2uyj1kulcHlJ:-LCp0LZk2uyj1kulcHlK:bafhed7) ([raw file](https://github.com/grpc/grpc-go/blob/a628f14e54172a5bdc9dda43db4237452f0a3685/credentials/credentials_util_go19.go#L35)):*
> ```Go
> }
> 
> func (c tlsConn) SyscallConn() (syscall.RawConn, error) {
> ```

Add a comment that this implements interface `syscall.Conn`

---


*Comments from [Reviewable](https://reviewable.io/reviews/grpc/grpc-go/1965)*
<!-- Sent from Reviewable.io -->

lyuxuan commented:
Review status: all files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[channelz/types_nonunixes.go, line 71 at r3](https://reviewable.io/reviews/grpc/grpc-go/1965#-LCdjIQl8RZWzvB90UvQ:-LCoMV9FDkCiZ5uWAK4S:beynskl) ([raw file](https://github.com/grpc/grpc-go/blob/a628f14e54172a5bdc9dda43db4237452f0a3685/channelz/types_nonunixes.go#L71)):*
<details><summary><i>Previously, menghanl (Menghan Li) wrote…</i></summary><blockquote>

This is a strong assumption that non-unix systems will have similar concepts with unix.

Looking at the proto def for socket option, the real data field is a `google.protobuf.Any`. We should hide the `Linger` details in the platform specific files, and add a method that generates the `Any` proto (move `MarshalAny` function calls from `service.go` to platform specific files).
</blockquote></details>

Discussed offline: to generate any proto, it will add pb dependency to channelz package, which will eventually lead to a cyclic import path. 

---

*[channelz/util_test.go, line 31 at r3](https://reviewable.io/reviews/grpc/grpc-go/1965#-LCoyDlF-bPD4sXmu54f:-LCp79NS0JuaacE9HkD1:bb7f2mf) ([raw file](https://github.com/grpc/grpc-go/blob/a628f14e54172a5bdc9dda43db4237452f0a3685/channelz/util_test.go#L31)):*
<details><summary><i>Previously, menghanl (Menghan Li) wrote…</i></summary><blockquote>

Remove this blank line.
</blockquote></details>

done

---

*[channelz/util_test.go, line 65 at r3](https://reviewable.io/reviews/grpc/grpc-go/1965#-LCp-Nff7CpsCG2DucD4:-LCp8AU12Wx2N3Platce:b-7ql1td) ([raw file](https://github.com/grpc/grpc-go/blob/a628f14e54172a5bdc9dda43db4237452f0a3685/channelz/util_test.go#L65)):*
<details><summary><i>Previously, menghanl (Menghan Li) wrote…</i></summary><blockquote>

something is wrong with this `startServer` function
</blockquote></details>

fixed

---

*[channelz/util_test.go, line 73 at r3](https://reviewable.io/reviews/grpc/grpc-go/1965#-LCoyef06tjGBxUMYg1U:-LCp8nxhC9XfHr46-UsO:bpem498) ([raw file](https://github.com/grpc/grpc-go/blob/a628f14e54172a5bdc9dda43db4237452f0a3685/channelz/util_test.go#L73)):*
<details><summary><i>Previously, menghanl (Menghan Li) wrote…</i></summary><blockquote>

The content of this struct will be modified by the syscall later, right?
Just leave it empty? Or use `new`.
</blockquote></details>

the value here is to set the fd, and we will read from the fd to verify the set and get functionality.

---

*[channelz/util_test.go, line 110 at r3](https://reviewable.io/reviews/grpc/grpc-go/1965#-LCozvUMCxPfZiSYKw_h:-LCp943H3tyF0jD0-5FQ:bb7f2mf) ([raw file](https://github.com/grpc/grpc-go/blob/a628f14e54172a5bdc9dda43db4237452f0a3685/channelz/util_test.go#L110)):*
<details><summary><i>Previously, menghanl (Menghan Li) wrote…</i></summary><blockquote>

defer
</blockquote></details>

done

---

*[channelz/service/service.go, line 38 at r2](https://reviewable.io/reviews/grpc/grpc-go/1965#-LBYOkKhCbOt1CcuAJQ5:-LCoSILjCV5Ov8LQ2JNK:b4ulm2e) ([raw file](https://github.com/grpc/grpc-go/blob/686bbac22514f2fb21aacbd96abfa3ef25e6faac/channelz/service/service.go#L38)):*
> ptypes.DurationProto
done

---

*[channelz/service/service.go, line 153 at r3](https://reviewable.io/reviews/grpc/grpc-go/1965#-LCp-cso9bN7DD_9izAJ:-LCp9xwR39-b5Hd8d0Wt:bb7f2mf) ([raw file](https://github.com/grpc/grpc-go/blob/a628f14e54172a5bdc9dda43db4237452f0a3685/channelz/service/service.go#L153)):*
<details><summary><i>Previously, menghanl (Menghan Li) wrote…</i></summary><blockquote>

Move this code around so you create one local variable and only set its field when error is not nil.
</blockquote></details>

done

---

*[channelz/service/service.go, line 168 at r3](https://reviewable.io/reviews/grpc/grpc-go/1965#-LCow4sY4raX9G6AX-mM:-LCpA-ub6DVSEhQlzE8X:bb7f2mf) ([raw file](https://github.com/grpc/grpc-go/blob/a628f14e54172a5bdc9dda43db4237452f0a3685/channelz/service/service.go#L168)):*
<details><summary><i>Previously, menghanl (Menghan Li) wrote…</i></summary><blockquote>

Move this and other code that accesses internal fields of SocketOptionData to _unix.go and _nonunix.go files. So you won't need the dummy non-unix types, either.
</blockquote></details>

done

---

*[channelz/service/service.go, line 168 at r3](https://reviewable.io/reviews/grpc/grpc-go/1965#-LCp1WHL5zNTsJSnhDBM:-LCpAXy9Dbb26WUx4DcC:bb7f2mf) ([raw file](https://github.com/grpc/grpc-go/blob/a628f14e54172a5bdc9dda43db4237452f0a3685/channelz/service/service.go#L168)):*
<details><summary><i>Previously, menghanl (Menghan Li) wrote…</i></summary><blockquote>

And put the fields of the struct to multiple lines.
</blockquote></details>

done

---

*[channelz/service/service_test.go, line 35 at r3](https://reviewable.io/reviews/grpc/grpc-go/1965#-LCp2V5R0AiCilGwflS1:-LCpAb6hB2OfmLvSCZVf:b-gwv4he) ([raw file](https://github.com/grpc/grpc-go/blob/a628f14e54172a5bdc9dda43db4237452f0a3685/channelz/service/service_test.go#L35)):*
<details><summary><i>Previously, menghanl (Menghan Li) wrote…</i></summary><blockquote>

channelzpb "google.golang.org/grpc/channelz/"
</blockquote></details>

I think this is obsolete, right?

---

*[channelz/service/service_test.go, line 41 at r3](https://reviewable.io/reviews/grpc/grpc-go/1965#-LCp4Sx79HuTsqYYmZsp:-LCpBBnf-C3t-NfKdpGJ:bb7f2mf) ([raw file](https://github.com/grpc/grpc-go/blob/a628f14e54172a5bdc9dda43db4237452f0a3685/channelz/service/service_test.go#L41)):*
<details><summary><i>Previously, menghanl (Menghan Li) wrote…</i></summary><blockquote>

Move this next to the definition of `OtherSecurityValue`, and add a comment that it's needed by `UnmarshalAny`
</blockquote></details>

done

---

*[channelz/service/service_test.go, line 180 at r3](https://reviewable.io/reviews/grpc/grpc-go/1965#-LCp10KD4-0G6fx0UjFu:-LCpDR42EQCA1Bw_w1Zl:bb7f2mf) ([raw file](https://github.com/grpc/grpc-go/blob/a628f14e54172a5bdc9dda43db4237452f0a3685/channelz/service/service_test.go#L180)):*
<details><summary><i>Previously, menghanl (Menghan Li) wrote…</i></summary><blockquote>

Similar to previous comments on making a function to convert duration, this and the following `1e3`, `1e9` could also be a function.
</blockquote></details>

done

---

*[credentials/credentials.go, line 237 at r3](https://reviewable.io/reviews/grpc/grpc-go/1965#-LCZt9nJ2PkgcB1nQUu9:-LCpDTmQ76xG1FAC658C:ba113nd) ([raw file](https://github.com/grpc/grpc-go/blob/a628f14e54172a5bdc9dda43db4237452f0a3685/credentials/credentials.go#L237)):*
<details><summary><i>Previously, menghanl (Menghan Li) wrote…</i></summary><blockquote>

The names for the interfaces are OK in channelz package, but inappropriate in credentials package.

Rename this or move it to channelz/subpackage
</blockquote></details>

rename to ExtraSecurityInfo. It's better to be in credentials package as people implementing credentials will notice it.

---

*[credentials/credentials_util_go19.go, line 35 at r3](https://reviewable.io/reviews/grpc/grpc-go/1965#-LCp0LZk2uyj1kulcHlJ:-LCpECNeDyBoEQLemNlO:bb7f2mf) ([raw file](https://github.com/grpc/grpc-go/blob/a628f14e54172a5bdc9dda43db4237452f0a3685/credentials/credentials_util_go19.go#L35)):*
<details><summary><i>Previously, menghanl (Menghan Li) wrote…</i></summary><blockquote>

Add a comment that this implements interface `syscall.Conn`
</blockquote></details>

done

---


*Comments from [Reviewable](https://reviewable.io/reviews/grpc/grpc-go/1965)*
<!-- Sent from Reviewable.io -->

menghanl commented:
Reviewed 10 of 13 files at r4.
Review status: 18 of 19 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[channelz/util_test.go, line 65 at r3](https://reviewable.io/reviews/grpc/grpc-go/1965#-LCp-Nff7CpsCG2DucD4:-LCpgH539_I2lmJZ-xi8:b-djk77g) ([raw file](https://github.com/grpc/grpc-go/blob/a628f14e54172a5bdc9dda43db4237452f0a3685/channelz/util_test.go#L65)):*
<details><summary><i>Previously, lyuxuan wrote…</i></summary><blockquote>

fixed
</blockquote></details>

But what's the purpose of function `startServer` and why the for loop plus the channel.

It seems to me what you actually want is just
```go
ln, _ := net.Listen(network, addr)
go func() { ln.Accept() } ()

conn, _ := net.Dial()

```

---


*Comments from [Reviewable](https://reviewable.io/reviews/grpc/grpc-go/1965)*
<!-- Sent from Reviewable.io -->

menghanl commented:
Reviewed 1 of 13 files at r4, 8 of 8 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[credentials/credentials.go, line 237 at r5](https://reviewable.io/reviews/grpc/grpc-go/1965#-LD7lRGWB1hCgiAUEE4-:-LD7lRGWB1hCgiAUEE40:b-p6ojig) ([raw file](https://github.com/grpc/grpc-go/blob/7814b524c1f85c6e2039607756d2cda3f268f1d2/credentials/credentials.go#L237)):*
> ```Go
> // ExtraSecurityInfo defines the interface that security protocols should implement in order
> // to provide security info to channelz.
> type ExtraSecurityInfo interface {
> ```

Nit: If there's no plan to use this and the following types for anything else, I would include the word "channelz" in all the names, like `ChannelzSecurityInfo`.

---


*Comments from [Reviewable](https://reviewable.io/reviews/grpc/grpc-go/1965)*
<!-- Sent from Reviewable.io -->

menghanl commented:
Breaks darwin builds. Revert in #2096.
